controllers/monitor.go

```diff
@@ -0,0 +1,169 @@
+package controllers
+
+import (
+	"context"
+	monitoringv1 "github.com/prometheus-operator/prometheus-operator/pkg/apis/monitoring/v1"
+	trustyaiopendatahubiov1alpha1 "github.com/trustyai-explainability/trustyai-service-operator/api/v1alpha1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+var isCentralServiceMonitorCreated = false
+
+// generateServiceMonitorSpecCentral generates the ServiceMonitor spec for central ServiceMonitor
+func generateServiceMonitorSpecCentral(deploymentNamespace string) *monitoringv1.ServiceMonitor {
+	serviceMonitor := &monitoringv1.ServiceMonitor{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      serviceMonitorName,
+			Namespace: deploymentNamespace,
+			Labels: map[string]string{
+				"modelmesh-service": "modelmesh-serving",
+			},
+		},
+		Spec: monitoringv1.ServiceMonitorSpec{
+			NamespaceSelector: monitoringv1.NamespaceSelector{
+				Any: true,
+			},
+			Endpoints: []monitoringv1.Endpoint{
+				{
+					Interval:    "4s",
+					Path:        "/q/metrics",
+					HonorLabels: true,
+					Scheme:      "http",
+					Params: map[string][]string{
+						"match[]": {
+							`{__name__= "trustyai_spd"}`,
+							`{__name__= "trustyai_dir"}`,
+						},
+					},
+					MetricRelabelConfigs: []*monitoringv1.RelabelConfig{
+						{
+							Action:       "keep",
+							Regex:        "trustyai_.*",
+							SourceLabels: []monitoringv1.LabelName{"__name__"},
+						},
+					},
+				},
+			},
+			Selector: metav1.LabelSelector{
+				MatchLabels: map[string]string{
+					"app.kubernetes.io/part-of": serviceType,
+				},
+			},
+		},
+	}
+	return serviceMonitor
+}
+
+// ensureCentralServiceMonitor ensures that the central ServiceMonitor is created
+func (r *TrustyAIServiceReconciler) ensureCentralServiceMonitor(ctx context.Context) error {
+	// Create only one central ServiceMonitor
+	if isCentralServiceMonitorCreated {
+		return nil
+	}
+
+	serviceMonitor := generateServiceMonitorSpecCentral(r.Namespace)
+
+	// Check if this ServiceMonitor already exists
+	found := &monitoringv1.ServiceMonitor{}
+	err := r.Get(ctx, types.NamespacedName{Name: serviceMonitor.Name, Namespace: serviceMonitor.Namespace}, found)
```

Comment:
Can you please clarify this part? As far as I see the `isCentralServiceMonitorCreated` is just a shortcut to avoid to check if the resource already exists.
Am I right?
I'm not sure if it is reliable: for example the controller can be restarted (so the variable set to `false` again) or even worst the `ServiceMonitor` CR can be manually cancelled and it should be recreated.

If it is only an optimization I would prefer to stay on the safe side and do the `GET` every time for now.
Wdyt?

Reply:
Removed the boolean flag. Central SM is not part of the reconcile loop.